File: python/Arrays&Hashing/valid_sudoku.py
from typing import List
from collections import defaultdict

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        """
        Approach:
        - Use three sets for each row, column, and 3x3 square to track seen numbers.
        - For each cell, if it's not empty, check if the number is already in the corresponding row, column, or square.
        - If so, return False. Otherwise, add the number to the sets.
        - If no duplicates are found, return True.

        Time Complexity: O(1), since the board size is fixed (9x9 = 81 cells).
        Space Complexity: O(1), since the sets store at most 9 numbers for each row, column, and square.
        """
        cols = defaultdict(set)
        rows = defaultdict(set)
        squares = defaultdict(set) # key: (row//3, col//3)
        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    continue
                if (board[r][c] in rows[r] or 
                    board[r][c] in cols[c] or 
                    board[r][c] in squares[r//3, c//3]):
                    return False
                cols[c].add(board[r][c])
                rows[r].add(board[r][c])
                squares[r//3, c//3].add(board[r][c])
        return True

# -------------------
# Test Cases
# -------------------
def test_valid_sudoku():
    sol = Solution()
    # Test case 1: Valid board
    board1 = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    assert sol.isValidSudoku(board1) == True
    # Test case 2: Invalid due to row
    board2 = [
        ["8","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    assert sol.isValidSudoku(board2) == False
    # Test case 3: Invalid due to column
    board3 = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        ["5",".",".",".","8",".",".","7","9"]
    ]
    assert sol.isValidSudoku(board3) == False
    # Test case 4: Invalid due to square
    board4 = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".","3","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]
    assert sol.isValidSudoku(board4) == False
    print("All test cases passed!")

if __name__ == "__main__":
    test_valid_sudoku() 

File: C++/Arrays&Hashing/valid_sudoku.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Validate if a Sudoku board is valid
     * 
     * Approach:
     * - Use hash sets to track numbers in each row, column, and 3x3 square
     * - Check each cell: if number exists in any set, board is invalid
     * - Otherwise add number to corresponding sets
     * 
     * Time Complexity: O(1) - board is fixed 9x9 size
     * Space Complexity: O(1) - sets store max 9 numbers each
     */
    bool isValidSudoku(vector<vector<char>>& board) {
        // Hash sets to store seen numbers for each row, column, and square
        vector<unordered_set<char>> rows(9);
        vector<unordered_set<char>> cols(9);
        vector<unordered_set<char>> squares(9);
        
        // Check each cell in the board
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                char curr = board[r][c];
                if (curr == '.') continue;
                
                // Calculate square index (0-8) from row and column
                int squ = (r/3) * 3 + (c/3);
                
                // Check if number already exists in any set
                if (rows[r].count(curr) || 
                    cols[c].count(curr) || 
                    squares[squ].count(curr)) {
                    return false;
                }
                
                // Add number to all three sets
                rows[r].insert(curr);
                cols[c].insert(curr);
                squares[squ].insert(curr);
            }
        }
        return true;
    }
};

/*
 * Detailed C++ Syntax Explanation:
 * 
 * 1. Vector of Sets:
 *    vector<unordered_set<char>> rows(9)
 *    - Creates 9 empty hash sets for rows/cols/squares
 *    - unordered_set provides O(1) lookup/insert
 * 
 * 2. Board Access:
 *    board[r][c]    // Access cell at row r, column c
 *    char curr      // Current cell value
 * 
 * 3. Set Operations:
 *    set.count(curr)   // Check if element exists (returns 0 or 1)
 *    set.insert(curr)  // Add element to set
 * 
 * 4. Square Index:
 *    (r/3) * 3 + (c/3)
 *    - Maps (r,c) to 0-8 index for squares
 *    - Integer division for block number
 */
